File: crates/voidx-agent/src/attachments.rs
//! Parse and materialize user file attachments.
//!
//! Ported from `src/voidx/agent/attachments.py`.
//! Handles: @path resolution, image base64 encoding, directory tree expansion.

use base64::Engine;
use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use std::path::{Path, PathBuf};

// pathdiff is used for computing relative paths between directories.
// If unavailable, we fall back to strip_prefix.

/// Image extensions that get base64-encoded as image parts.
pub const IMAGE_EXTENSIONS: &[&str] = &[".png", ".jpg", ".jpeg", ".gif", ".webp"];

/// Max size for text attachments (bytes).
pub const MAX_TEXT_ATTACHMENT_BYTES: u64 = 200_000;
/// Max size for image attachments (bytes).
pub const MAX_IMAGE_ATTACHMENT_BYTES: u64 = 5_000_000;
/// Max items in a directory listing.
pub const MAX_DIR_LISTING_ITEMS: usize = 500;

/// Directories to skip when building directory trees.
pub const DIR_TREE_SKIP: &[&str] = &[
    "__pycache__",
    ".git",
    ".hg",
    ".svn",
    "node_modules",
    ".venv",
    "venv",
    "dist",
    "build",
    ".pytest_cache",
    ".mypy_cache",
    "target",
];

/// An attachment parsed from user input.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Attachment {
    pub path: PathBuf,
    pub rel_path: String,
    pub kind: AttachmentKind,
    pub mime_type: String,
    pub size: u64,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum AttachmentKind {
    File,
    Image,
    Dir,
}

/// The processed user message payload.
#[derive(Debug, Clone)]
pub struct UserMessagePayload {
    /// Original text with @path tokens.
    pub raw_text: String,
    /// Text with @path tokens replaced by section markers.
    pub clean_text: String,
    /// Text suitable for display (with [image-x] markers).
    pub display_text: String,
    /// Short title for UI.
    pub title_text: String,
    /// Final content — either a plain string or structured content with image parts.
    pub content: MessageContent,
    /// Parsed attachments.
    pub attachments: Vec<Attachment>,
    /// Warnings generated during parsing.
    pub warnings: Vec<String>,
}

/// Message content can be plain text or structured (with image parts).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum MessageContent {
    Text(String),
    Structured(Vec<ContentPart>),
}

/// A single part of structured content.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum ContentPart {
    #[serde(rename = "text")]
    Text { text: String },
    #[serde(rename = "image_url")]
    Image { image_url: ImageUrl },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ImageUrl {
    pub url: String,
}

/// Parse user input and resolve all @path attachments.
pub fn build_user_message_payload(
    user_text: &str,
    workspace: &Path,
) -> UserMessagePayload {
    let workspace = match workspace.canonicalize() {
        Ok(p) => p,
        Err(_) => workspace.to_path_buf(),
    };

    let tokens = extract_attachment_tokens(user_text);
    let mut clean_parts: Vec<String> = Vec::new();
    let mut display_parts: Vec<String> = Vec::new();
    let mut attachments: Vec<Attachment> = Vec::new();
    let mut warnings: Vec<String> = Vec::new();
    let mut image_parts: Vec<ContentPart> = Vec::new();
    let mut text_sections: Vec<ContentPart> = Vec::new();

    let mut last_end = 0usize;

    for token in &tokens {
        // Add text before this token
        if token.start > last_end {
            let before = &user_text[last_end..token.start];
            if !before.is_empty() {
                clean_parts.push(before.to_string());
                display_parts.push(before.to_string());
                text_sections.push(ContentPart::Text {
                    text: before.to_string(),
                });
            }
        }

        let raw_path = &token.raw_path;
        let resolved = resolve_workspace_path(&workspace, raw_path);

        match resolved {
            None => {
                warnings.push(format!("Attachment skipped outside workspace: {raw_path}"));
                clean_parts.push(raw_path.clone());
                display_parts.push(raw_path.clone());
                text_sections.push(ContentPart::Text {
                    text: raw_path.clone(),
                });
            }
            Some(path) if !path.exists() => {
                warnings.push(format!("Attachment not found: {raw_path}"));
                clean_parts.push(raw_path.clone());
                display_parts.push(raw_path.clone());
                text_sections.push(ContentPart::Text {
                    text: raw_path.clone(),
                });
            }
            Some(path) => {
                let is_dir = path.is_dir();
                let rel_path = pathdiff::diff_paths(&path, &workspace)
                    .unwrap_or_else(|| path.clone())
                    .to_string_lossy()
                    .to_string();

                let mime_type = guess_mime_type(&path);
                let size = path.metadata().map(|m| m.len()).unwrap_or(0);

                let kind = if is_dir {
                    AttachmentKind::Dir
                } else if is_image_path(&path) {
                    AttachmentKind::Image
                } else {
                    AttachmentKind::File
                };

                let attachment = Attachment {
                    path: path.clone(),
                    rel_path: rel_path.clone(),
                    kind,
                    mime_type: mime_type.clone(),
                    size,
                };
                attachments.push(attachment);

                match kind {
                    AttachmentKind::Image => {
                        if size > MAX_IMAGE_ATTACHMENT_BYTES {
                            warnings.push(format!(
                                "Image skipped because it is too large: {rel_path}"
                            ));
                        } else {
                            let section = format!("\nAttached image: {rel_path}\n");
                            clean_parts.push(section.clone());
                            display_parts.push(format!("[image-{rel_path}]"));
                            if let Some(image_part) = encode_image(&path, &mime_type) {
                                image_parts.push(image_part);
                            }
                        }
                    }
                    AttachmentKind::Dir => {
                        let (section, warning) = directory_section(&path, &workspace);
                        if let Some(w) = warning {
                            warnings.push(w);
                        }
                        clean_parts.push(section.clone());
                        display_parts.push(format!("[dir-{rel_path}]"));
                        text_sections.push(ContentPart::Text { text: section });
                    }
                    AttachmentKind::File => {
                        if size > MAX_TEXT_ATTACHMENT_BYTES {
                            warnings.push(format!(
                                "File skipped because it is too large: {rel_path} ({size} bytes)"
                            ));
                        } else {
                            let content = std::fs::read_to_string(&path).unwrap_or_default();
                            let section = format!(
                                "\nAttached file: {rel_path}\n```\n{content}\n```\n"
                            );
                            clean_parts.push(section.clone());
                            display_parts.push(format!("[file-{rel_path}]"));
                            text_sections.push(ContentPart::Text { text: section });
                        }
                    }
                }
            }
        }

        last_end = token.end;
    }

    // Add remaining text after last token
    if last_end < user_text.len() {
        let remaining = &user_text[last_end..];
        if !remaining.is_empty() {
            clean_parts.push(remaining.to_string());
            display_parts.push(remaining.to_string());
            text_sections.push(ContentPart::Text {
                text: remaining.to_string(),
            });
        }
    }

    let clean_text = clean_parts.join("");
    let display_text = display_parts.join("");
    let title_text = make_title(&display_text);

    // Build final content
    let content = if image_parts.is_empty() {
        MessageContent::Text(clean_text.clone())
    } else {
        let mut parts: Vec<ContentPart> = text_sections;
        parts.extend(image_parts);
        MessageContent::Structured(parts)
    };

    UserMessagePayload {
        raw_text: user_text.to_string(),
        clean_text,
        display_text,
        title_text,
        content,
        attachments,
        warnings,
    }
}

// ── Attachment token extraction ──────────────────────────────────────────

/// A parsed @path token from user text.
#[derive(Debug)]
struct AttachmentToken {
    start: usize,
    end: usize,
    raw_path: String,
}

/// Extract all @path tokens from user text.
///
/// Supports: @path/to/file, @"path with spaces", [image-name]
///
/// All indices are byte offsets into `text`, safe for slicing.
fn extract_attachment_tokens(text: &str) -> Vec<AttachmentToken> {
    let mut tokens = Vec::new();
    let bytes = text.as_bytes();
    let mut i = 0usize; // byte index

    while i < bytes.len() {
        // @path or @"quoted path"
        if bytes[i] == b'@' {
            let start = i;
            i += 1;

            if i < bytes.len() && bytes[i] == b'"' {
                // Quoted path
                i += 1;
                let path_start = i;
                while i < bytes.len() && bytes[i] != b'"' {
                    i += 1;
                }
                let raw_path = text[path_start..i].to_string();
                if i < bytes.len() {
                    i += 1; // skip closing quote
                }
                tokens.push(AttachmentToken {
                    start,
                    end: i,
                    raw_path,
                });
            } else {
                // Unquoted path
                let path_start = i;
                while i < bytes.len() && !bytes[i].is_ascii_whitespace() {
                    i += 1;
                }
                let raw_path = text[path_start..i].to_string();
                if !raw_path.is_empty() {
                    tokens.push(AttachmentToken {
                        start,
                        end: i,
                        raw_path,
                    });
                }
            }
            continue;
        }

        // [image-name]
        if bytes[i] == b'[' && i + 1 < bytes.len() {
            let bracket_start = i;
            i += 1;
            if text[i..].starts_with("image-") {
                i += 6; // skip "image-"
                let name_start = i;
                while i < bytes.len() && bytes[i] != b']' {
                    i += 1;
                }
                if i < bytes.len() {
                    let name = text[name_start..i].to_string();
                    i += 1; // skip ]
                    tokens.push(AttachmentToken {
                        start: bracket_start,
                        end: i,
                        raw_path: format!(":image:{name}"),
                    });
                    continue;
                }
            }
            i = bracket_start + 1;
            continue;
        }

        i += 1;
    }

    tokens
}

// ── Path resolution ──────────────────────────────────────────────────────

/// Resolve a path relative to the workspace, ensuring it stays inside.
fn resolve_workspace_path(workspace: &Path, raw_path: &str) -> Option<PathBuf> {
    // Handle :image: prefix
    if let Some(stem) = raw_path.strip_prefix(":image:") {
        // Try to find the image in the workspace
        for ext in IMAGE_EXTENSIONS {
            let candidate = workspace.join(format!("{stem}{ext}"));
            if candidate.exists() {
                return Some(candidate);
            }
        }
        // Try the stem as a full path
        let candidate = workspace.join(stem);
        if candidate.exists() {
            return Some(candidate);
        }
        return None;
    }

    let resolved = workspace.join(raw_path);

    // Verify it's inside workspace
    match resolved.canonicalize() {
        Ok(canonical) => {
            if let Ok(ws_canonical) = workspace.canonicalize() {
                if canonical.starts_with(&ws_canonical) {
                    return Some(canonical);
                }
            }
            // Fallback: if canonicalize fails for workspace, allow it
            Some(canonical)
        }
        Err(_) => {
            // Path doesn't exist yet — still allow if it's under workspace
            if resolved.starts_with(workspace) {
                Some(resolved)
            } else {
                None
            }
        }
    }
}

/// Check if a path has an image extension.
fn is_image_path(path: &Path) -> bool {
    path.extension()
        .and_then(|e| e.to_str())
        .map(|e| IMAGE_EXTENSIONS.contains(&e.to_lowercase().as_str()))
        .unwrap_or(false)
}

/// Guess MIME type from file extension.
fn guess_mime_type(path: &Path) -> String {
    match path.extension().and_then(|e| e.to_str()).map(|e| e.to_lowercase()) {
        Some(e) => match e.as_str() {
            "png" => "image/png".to_string(),
            "jpg" | "jpeg" => "image/jpeg".to_string(),
            "gif" => "image/gif".to_string(),
            "webp" => "image/webp".to_string(),
            "svg" => "image/svg+xml".to_string(),
            "json" => "application/json".to_string(),
            "xml" => "application/xml".to_string(),
            "html" => "text/html".to_string(),
            "md" => "text/markdown".to_string(),
            "txt" => "text/plain".to_string(),
            "rs" => "text/rust".to_string(),
            "py" => "text/x-python".to_string(),
            "ts" | "tsx" => "text/typescript".to_string(),
            "js" | "jsx" => "text/javascript".to_string(),
            _ => "application/octet-stream".to_string(),
        },
        None => "application/octet-stream".to_string(),
    }
}

/// Encode an image file as a base64 data URL.
fn encode_image(path: &Path, mime_type: &str) -> Option<ContentPart> {
    let data = std::fs::read(path).ok()?;
    let b64 = base64::engine::general_purpose::STANDARD.encode(&data);
    Some(ContentPart::Image {
        image_url: ImageUrl {
            url: format!("data:{mime_type};base64,{b64}"),
        },
    })
}

/// Build a directory tree section for an attached directory.
fn directory_section(dir_path: &Path, workspace: &Path) -> (String, Option<String>) {
    let skip: HashSet<&str> = DIR_TREE_SKIP.iter().copied().collect();
    let mut entries = Vec::new();
    let mut warning = None;

    // Simple directory listing
    collect_dir_entries(dir_path, &skip, &mut entries, 0, &mut warning);

    let rel = pathdiff::diff_paths(dir_path, workspace)
        .unwrap_or_else(|| dir_path.to_path_buf());
    let rel_str = rel.to_string_lossy();

    let mut section = format!("\nAttached directory: {rel_str}\n```\n");
    for entry in &entries {
        section.push_str(entry);
        section.push('\n');
    }
    section.push_str("```\n");

    (section, warning)
}

/// Recursively collect directory entries for display.
fn collect_dir_entries(
    dir: &Path,
    skip: &HashSet<&str>,
    entries: &mut Vec<String>,
    depth: usize,
    warning: &mut Option<String>,
) {
    if entries.len() >= MAX_DIR_LISTING_ITEMS {
        if warning.is_none() {
            *warning = Some(format!(
                "Directory listing truncated at {} items",
                MAX_DIR_LISTING_ITEMS
            ));
        }
        return;
    }

    let mut subdirs: Vec<PathBuf> = Vec::new();
    let mut files: Vec<String> = Vec::new();

    if let Ok(rd) = std::fs::read_dir(dir) {
        for entry in rd.flatten() {
            let name = entry.file_name().to_string_lossy().to_string();
            if skip.contains(name.as_str()) || name.starts_with('.') {
                continue;
            }

            let path = entry.path();
            if path.is_dir() {
                subdirs.push(path);
            } else {
                files.push(name);
            }
        }
    }

    // Sort for deterministic output
    subdirs.sort();
    files.sort();

    let indent = "  ".repeat(depth);

    for subdir in &subdirs {
        let name = subdir.file_name().unwrap_or_default().to_string_lossy();
        entries.push(format!("{indent}{name}/"));
        collect_dir_entries(subdir, skip, entries, depth + 1, warning);
    }

    for file in &files {
        entries.push(format!("{indent}{file}"));
    }
}

/// Make a short title from display text.
fn make_title(text: &str) -> String {
    let first_line = text.lines().next().unwrap_or(text);
    if first_line.len() > 80 {
        format!("{}...", &first_line[..77])
    } else {
        first_line.to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_extract_tokens_simple() {
        let tokens = extract_attachment_tokens("look at @src/main.rs please");
        assert_eq!(tokens.len(), 1);
        assert_eq!(tokens[0].raw_path, "src/main.rs");
    }

    #[test]
    fn test_extract_tokens_quoted() {
        let tokens = extract_attachment_tokens("look at @\"path with spaces.rs\"");
        assert_eq!(tokens.len(), 1);
        assert_eq!(tokens[0].raw_path, "path with spaces.rs");
    }

    #[test]
    fn test_extract_tokens_image() {
        let tokens = extract_attachment_tokens("see [image-photo]");
        assert_eq!(tokens.len(), 1);
        assert_eq!(tokens[0].raw_path, ":image:photo");
    }

    #[test]
    fn test_extract_tokens_multiple() {
        let tokens = extract_attachment_tokens("@a.txt and @b.rs");
        assert_eq!(tokens.len(), 2);
        assert_eq!(tokens[0].raw_path, "a.txt");
        assert_eq!(tokens[1].raw_path, "b.rs");
    }

    #[test]
    fn test_is_image_path() {
        assert!(is_image_path(Path::new("photo.png")));
        assert!(is_image_path(Path::new("photo.JPG")));
        assert!(!is_image_path(Path::new("code.rs")));
    }

    #[test]
    fn test_guess_mime_type() {
        assert_eq!(guess_mime_type(Path::new("f.png")), "image/png");
        assert_eq!(guess_mime_type(Path::new("f.rs")), "text/rust");
        assert_eq!(guess_mime_type(Path::new("f.json")), "application/json");
    }

    #[test]
    fn test_sanitize_empty_input() {
        let payload = build_user_message_payload("hello world", Path::new("/tmp"));
        assert_eq!(payload.raw_text, "hello world");
        assert!(payload.attachments.is_empty());
        assert!(payload.warnings.is_empty());
    }

    #[test]
    fn test_make_title() {
        assert_eq!(make_title("short"), "short");
        let long = "x".repeat(100);
        let title = make_title(&long);
        assert!(title.ends_with("..."));
        assert_eq!(title.len(), 80);
    }
}

File: src/voidx/agent/attachments.py
"""Parse and materialize user file attachments."""

from __future__ import annotations

import base64
import json
import mimetypes
import re
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from voidx.runtime.attachments import MAX_IMAGE_ATTACHMENT_BYTES

IMAGE_EXTENSIONS = {".png", ".jpg", ".jpeg", ".gif", ".webp"}
MAX_TEXT_ATTACHMENT_BYTES = 200_000
MAX_DIR_LISTING_ITEMS = 500
_DIR_TREE_SKIP = {"__pycache__", ".git", ".hg", ".svn", "node_modules", ".venv", "venv", "dist", "build", ".pytest_cache", ".mypy_cache"}
_ATTACHMENT_RE = re.compile(r'(?<!\S)(?:@(?:"([^"]+)"|(\S+))|\[image-([^\]]+)\])')
_CLIPBOARD_ATTACHMENT_DIR = ".voidx/attachments"


@dataclass(frozen=True)
class Attachment:
    path: Path
    rel_path: str
    kind: str
    mime_type: str
    size: int


@dataclass(frozen=True)
class PathReference:
    raw_path: str
    display_path: str
    token_span: tuple[int, int]


@dataclass
class UserMessagePayload:
    raw_text: str
    clean_text: str
    display_text: str
    title_text: str
    content: str | list[dict[str, Any]]
    content_format: str
    attachments: list[Attachment] = field(default_factory=list)
    path_references: list[PathReference] = field(default_factory=list)
    warnings: list[str] = field(default_factory=list)


def parse_structured_content(content: str, content_format: str) -> str | list[dict[str, Any]]:
    if content_format != "structured":
        return content
    try:
        parsed = json.loads(content)
    except Exception:
        return content
    return parsed if isinstance(parsed, list) else content


def serialize_message_content(content: str | list[dict[str, Any]]) -> tuple[str, str]:
    if isinstance(content, list):
        return json.dumps(content, ensure_ascii=False), "structured"
    return content, "text"


def build_user_message_payload(
    user_text: str,
    workspace: str,
    *,
    text_prefix: str = "",
    extra_removed_spans: list[tuple[int, int]] | None = None,
) -> UserMessagePayload:
    workspace_path = Path(workspace).resolve()
    tokens = _attachment_tokens(user_text)
    removed_spans: list[tuple[int, int]] = []
    attachments: list[Attachment] = []
    path_references: list[PathReference] = []
    warnings: list[str] = []
    text_sections: list[str] = []
    image_parts: list[dict[str, Any]] = []

    for start, end, raw_path in tokens:
        if not raw_path.startswith(":image:"):
            removed_spans.append((start, end))
            path_references.append(
                PathReference(
                    raw_path=raw_path,
                    display_path=_display_reference_path(workspace_path, raw_path),
                    token_span=(start, end),
                )
            )
            continue

        stem = raw_path[len(":image:"):]
        resolved = _resolve_image_stem(workspace_path, stem)
        display_label = f"[image-{stem}]"
        if resolved is None:
            warnings.append(f"Image attachment not found: {display_label}")
            continue
        if not resolved.exists():
            warnings.append(f"Attachment not found: {display_label}")
            continue
        if not resolved.is_file() and not resolved.is_dir():
            warnings.append(f"Attachment is not a file or directory: {display_label}")
            continue
        is_dir = resolved.is_dir()
        removed_spans.append((start, end))
        rel_path = _relative_path(workspace_path, resolved)
        mime_type = mimetypes.guess_type(str(resolved))[0] or "application/octet-stream"
        size = resolved.stat().st_size
        kind = "dir" if is_dir else ("image" if is_image_path(resolved) else "file")
        attachment = Attachment(resolved, rel_path, kind, mime_type, size)
        attachments.append(attachment)

        if kind == "image":
            if size > MAX_IMAGE_ATTACHMENT_BYTES:
                warnings.append(f"Image skipped because it is too large: {rel_path}")
                continue
            image_parts.append(_image_part(resolved, mime_type))
            continue

        if kind == "dir":
            section, warning = _directory_section(attachment)
            if warning:
                warnings.append(warning)
            if section:
                text_sections.append(section)
            continue

        section, warning = _text_file_section(attachment)
        if warning:
            warnings.append(warning)
        if section:
            text_sections.append(section)

    display_clean_text = _normalize_text(_remove_spans(user_text, removed_spans))
    if extra_removed_spans:
        removed_spans.extend(extra_removed_spans)
    clean_text = _normalize_text(_remove_spans(user_text, removed_spans))
    text_content = _build_text_content(clean_text, attachments, text_sections, path_references)
    if text_prefix.strip():
        text_content = _prefix_text_content(text_prefix.strip(), text_content)
    content: str | list[dict[str, Any]]
    content_format = "text"
    if image_parts:
        content = [{"type": "text", "text": text_content}, *image_parts]
        content_format = "structured"
    else:
        content = text_content

    display_text = _display_text(display_clean_text, attachments, path_references)
    title_text = clean_text or (
        f"Referenced {path_references[0].display_path}"
        if path_references
        else (f"Attached {attachments[0].rel_path}" if attachments else user_text)
    )
    return UserMessagePayload(
        raw_text=user_text,
        clean_text=clean_text,
        display_text=display_text,
        title_text=title_text,
        content=content,
        content_format=content_format,
        attachments=attachments,
        path_references=path_references,
        warnings=warnings,
    )


def is_image_path(path: Path | str) -> bool:
    return Path(path).suffix.lower() in IMAGE_EXTENSIONS


_PASTED_RE = re.compile(r"<pasted>\n.*?\n</pasted>", re.DOTALL)


def _pasted_spans(text: str) -> list[tuple[int, int]]:
    """Return (start, end) spans of all <pasted>...</pasted> blocks."""
    return [(m.start(), m.end()) for m in _PASTED_RE.finditer(text)]


def _attachment_tokens(text: str) -> list[tuple[int, int, str]]:
    tokens: list[tuple[int, int, str]] = []
    excluded = _pasted_spans(text)
    for match in _ATTACHMENT_RE.finditer(text):
        # Skip @-references and [image-] tokens inside <pasted> blocks —
        # pasted content is quoted material (e.g. code with decorators),
        # not user-authored attachment references.
        if any(start <= match.start() < end for start, end in excluded):
            continue
        image_stem = match.group(3)
        if image_stem:
            raw_path = f":image:{image_stem}"
        else:
            raw_path = match.group(1) or match.group(2)
        if raw_path:
            tokens.append((match.start(), match.end(), raw_path))
    return tokens


def _remove_spans(text: str, spans: list[tuple[int, int]]) -> str:
    if not spans:
        return text
    result: list[str] = []
    last = 0
    for start, end in sorted(spans):
        if start < last:
            continue
        result.append(text[last:start])
        result.append(" ")
        last = end
    result.append(text[last:])
    return "".join(result)


def _display_reference_path(workspace: Path, raw_path: str) -> str:
    candidate = Path(raw_path).expanduser()
    try:
        resolved = candidate.resolve() if candidate.is_absolute() else (workspace / candidate).resolve()
    except (OSError, ValueError):
        return raw_path
    return _relative_path(workspace, resolved)


def _prefix_text_content(prefix: str, text: str) -> str:
    text = text.strip()
    if not text:
        return prefix
    return f"{prefix}\n\n{text}"




def _resolve_image_stem(workspace: Path, stem: str) -> Path | None:
    target_dir = workspace / _CLIPBOARD_ATTACHMENT_DIR
    if not target_dir.is_dir():
        return None
    matches = sorted(target_dir.glob(f"{stem}.*"))
    return matches[0] if matches else None


def _relative_path(workspace: Path, path: Path) -> str:
    try:
        return path.relative_to(workspace).as_posix()
    except ValueError:
        return path.as_posix()


def _image_part(path: Path, mime_type: str) -> dict[str, Any]:
    encoded = base64.b64encode(path.read_bytes()).decode("ascii")
    return {
        "type": "image_url",
        "image_url": {"url": f"data:{mime_type};base64,{encoded}"},
    }


def _text_file_section(attachment: Attachment) -> tuple[str, str]:
    raw = attachment.path.read_bytes()
    truncated = len(raw) > MAX_TEXT_ATTACHMENT_BYTES
    if truncated:
        raw = raw[:MAX_TEXT_ATTACHMENT_BYTES]
    if b"\x00" in raw:
        return (
            f"Attached binary file: {attachment.rel_path} ({attachment.mime_type}, {attachment.size} bytes).",
            "",
        )
    try:
        text = raw.decode("utf-8")
    except UnicodeDecodeError:
        text = raw.decode("utf-8", errors="replace")
    suffix = ""
    warning = ""
    if truncated:
        omitted = attachment.size - MAX_TEXT_ATTACHMENT_BYTES
        suffix = f"\n\n[Attachment truncated: omitted {omitted} bytes]"
        warning = f"Attachment truncated: {attachment.rel_path}"
    lang = _language_from_path(attachment.rel_path)
    return f"Attached file: {attachment.rel_path}\n```{lang}\n{text}{suffix}\n```", warning


def _directory_section(attachment: Attachment) -> tuple[str, str]:
    listing: list[str] = []
    try:
        _walk_dir_tree(attachment.path, "", listing, depth=0, max_depth=3)
    except (OSError, PermissionError) as exc:
        return "", f"Cannot read directory {attachment.rel_path}: {exc}"
    tree_text = "\n".join(listing) if listing else "(empty)"
    label = f"Attached directory: {attachment.rel_path}"
    if listing and listing[-1].startswith("..."):
        item_count = sum(1 for line in listing if not line.endswith("..."))
        label += f" (showing first {item_count} items)"
    return f"{label}\n{tree_text}", ""


def _walk_dir_tree(root: Path, prefix: str, listing: list[str], depth: int, max_depth: int) -> None:
    if depth >= max_depth or len(listing) >= MAX_DIR_LISTING_ITEMS:
        if not listing or not listing[-1].startswith("..."):
            listing.append(f"{prefix}...")
        return
    try:
        entries = sorted(root.iterdir(), key=lambda p: (not p.is_dir(), p.name.lower()))
    except (OSError, PermissionError):
        return
    for i, entry in enumerate(entries):
        if entry.name.startswith("."):
            continue
        if entry.is_dir() and entry.name in _DIR_TREE_SKIP:
            continue
        if len(listing) >= MAX_DIR_LISTING_ITEMS:
            if not listing[-1].startswith("..."):
                listing.append(f"{prefix}...")
            return
        is_last = i == len(entries) - 1
        connector = "└── " if is_last else "├── "
        listing.append(f"{prefix}{connector}{entry.name}")
        if entry.is_dir():
            ext_prefix = "    " if is_last else "│   "
            _walk_dir_tree(entry, prefix + ext_prefix, listing, depth + 1, max_depth)


def _language_from_path(path: str) -> str:
    suffix = Path(path).suffix.lower().lstrip(".")
    mapping = {
        "py": "python",
        "js": "javascript",
        "ts": "typescript",
        "tsx": "tsx",
        "jsx": "jsx",
        "json": "json",
        "md": "markdown",
        "sh": "bash",
        "yml": "yaml",
        "yaml": "yaml",
        "html": "html",
        "css": "css",
    }
    return mapping.get(suffix, suffix)


def _build_text_content(
    clean_text: str,
    attachments: list[Attachment],
    text_sections: list[str],
    path_references: list[PathReference],
) -> str:
    parts: list[str] = []
    if clean_text:
        parts.append(clean_text)
    if path_references:
        reference_lines = [f"- {item.display_path}" for item in path_references]
        parts.append("Referenced paths:\n" + "\n".join(reference_lines))
    if attachments:
        image_lines = [
            f"- {item.rel_path} ({item.mime_type}, {item.size} bytes)"
            for item in attachments
            if item.kind == "image"
        ]
        if image_lines:
            parts.append("Attached images:\n" + "\n".join(image_lines))
    parts.extend(text_sections)
    if parts:
        return "\n\n".join(parts)
    return "Please review the referenced path."


def _display_text(clean_text: str, attachments: list[Attachment], path_references: list[PathReference]) -> str:
    if path_references:
        names = ", ".join(item.display_path for item in path_references[:3])
        if len(path_references) > 3:
            names += f", +{len(path_references) - 3} more"
        base = clean_text or "Referenced paths"
        return f"{base}\n[references: {names}]"
    if not attachments:
        return clean_text
    names = ", ".join(item.rel_path for item in attachments[:3])
    if len(attachments) > 3:
        names += f", +{len(attachments) - 3} more"
    base = clean_text or "Attached files"
    return f"{base}\n[attachments: {names}]"

def _normalize_text(text: str) -> str:
    return re.sub(r"[ \t]+", " ", text).strip()

